cmd/snap-seccomp/main_test.go

```diff
@@ -47,147 +42,224 @@ import (
-
-	r, err := os.Open(p)
-	if err != nil {
-		return nil, err
+var seccompBpfLoaderContent = []byte(`
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <fcntl.h>
+#include <sys/prctl.h>
+
+#include <linux/filter.h>
+#include <linux/seccomp.h>
+
+#define MAX_BPF_SIZE 32 * 1024
+
+int sc_apply_seccomp_bpf(const char *profile_path)
+{
+	unsigned char bpf[MAX_BPF_SIZE + 1]; // account for EOF
+	FILE *fp;
+	fp = fopen(profile_path, "rb");
+	if (fp == NULL) {
+		fprintf(stderr, "cannot read %s\n", profile_path);
+		exit(1);
 	}
-	defer r.Close()
 
-	for {
-		err = binary.Read(r, nativeEndian(), &rawOp)
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			return nil, err
-		}
-		ops = append(ops, rawOp.Disassemble())
+	// set 'size' to '1; to get bytes transferred
+	size_t num_read = fread(bpf, 1, sizeof(bpf), fp);
+
+	if (ferror(fp) != 0) {
+		perror("fread()");
+		exit(1);
+	} else if (feof(fp) == 0) {
+		fprintf(stderr, "file too big\n");
+		exit(1);
+	}
+	fclose(fp);
+
+	struct sock_fprog prog = {
+		.len = num_read / sizeof(struct sock_filter),
+		.filter = (struct sock_filter *)bpf,
+	};
+
+	// Set NNP to allow loading seccomp policy into the kernel without
+	// root
+	if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)) {
+		perror("prctl(PR_NO_NEW_PRIVS, 1, 0, 0, 0)");
+		exit(1);
 	}
 
-	return ops, nil
+	if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &prog)) {
+		perror("prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, ...) failed");
+		exit(1);
+	}
+	return 0;
 }
 
-func parseBpfInput(s string) (*main.SeccompData, error) {
-	// syscall;arch;arg1,arg2...
-	l := strings.Split(s, ";")
-
-	var scmpArch seccomp.ScmpArch
-	if len(l) < 2 || l[1] == "native" {
-		scmpArch = main.UbuntuArchToScmpArch(arch.UbuntuArchitecture())
-	} else {
-		scmpArch = main.UbuntuArchToScmpArch(l[1])
+int main(int argc, char *argv[])
+{
+	int rc = 0;
+	if (argc < 2) {
+		fprintf(stderr, "Usage: %s <bpf file> [prog ...]\n", argv[0]);
+		return 1;
 	}
 
-	sc, err := seccomp.GetSyscallFromNameByArch(l[0], scmpArch)
-	if err != nil {
-		return nil, err
-	}
-	// libseccomp may return negative numbers here for syscalls that
-	// are "special" for some reason. there is no "official" way to
-	// resolve them using the API to the real number. this is why
-	// we workaround there.
-	if sc < 0 {
-		/* -101 is __PNR_socket */
-		if sc == -101 && scmpArch == seccomp.ArchX86 {
-			sc = 359 /* see src/arch-x86.c socket */
-		} else if sc == -101 && scmpArch == seccomp.ArchS390X {
-			sc = 359 /* see src/arch-s390x.c socket */
-		} else if sc == -10165 && scmpArch == seccomp.ArchARM64 {
-			// -10165 is mknod on aarch64 and it is translated
-			// to mknodat. for our simulation -10165 is fine
-			// though
-		} else {
-			panic(fmt.Sprintf("cannot resolve syscall %v for arch %v, got %v", l[0], l[1], sc))
-		}
-	}
+	rc = sc_apply_seccomp_bpf(argv[1]);
+	if (rc || argc == 2)
```

Comment:
``sc_apply_seccomp_bpf`` doesn't seem to be reporting any errors by return value, it exits right away or returns 0. I think returning instead of exiting makes it more friendly for testing, but not insisting on that. In any case, this conditional seems inconsistent with what the function is doing.

Reply:
Thanks, I addressed this now.

Reply:
Thanks! Is there a particular reason you return negative value from the function (only to return -rc from main)?